crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -0,0 +1,477 @@
+    /// for serialization between runs.
+    ///
+    /// No particular structure is guaranteed for the returned bits.
+    pub fn to_bits(self) -> u64 {
+        u64::from(self.generation) << 32 | u64::from(self.id)
+    }
+
+    /// Reconstruct an `Entity` previously destructured with `to_bits`
+    ///
+    /// Only useful when applied to results from `to_bits` in the same instance of an application.
+    pub fn from_bits(bits: u64) -> Self {
+        Self {
+            generation: (bits >> 32) as u32,
+            id: bits as u32,
+        }
+    }
+
+    /// Return a transiently unique identifier
+    ///
+    /// No two simultaneously-live entities share the same ID, but dead entities' IDs may collide
+    /// with both live and dead entities. Useful for compactly representing entities within a
+    /// specific snapshot of the world, such as when serializing.
+    #[inline]
+    pub fn id(self) -> u32 {
+        self.id
+    }
+
+    /// Returns the generation of this Entity's id. The generation is incremented each time an entity with
+    /// a given id is despawned. This serves as a "count" of the number of times a given id has been reused
+    /// (id, generation) pairs uniquely identify a given Entity.
+    #[inline]
+    pub fn generation(self) -> u32 {
+        self.generation
+    }
+}
+
+impl fmt::Debug for Entity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}v{}", self.id, self.generation)
+    }
+}
+
+impl SparseSetIndex for Entity {
+    fn sparse_set_index(&self) -> usize {
+        self.id() as usize
+    }
+
+    fn get_sparse_set_index(value: usize) -> Self {
+        Entity::new(value as u32)
+    }
+}
+
+/// An iterator returning a sequence of Entity values from `Entities::reserve_entities`.
+pub struct ReserveEntitiesIterator<'a> {
+    // Metas, so we can recover the current generation for anything in the freelist.
+    meta: &'a [EntityMeta],
+
+    // Reserved IDs formerly in the freelist to hand out.
+    id_iter: std::slice::Iter<'a, u32>,
+
+    // New Entity IDs to hand out, outside the range of meta.len().
+    id_range: std::ops::Range<u32>,
+}
+
+impl<'a> Iterator for ReserveEntitiesIterator<'a> {
+    type Item = Entity;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.id_iter
+            .next()
+            .map(|&id| Entity {
+                generation: self.meta[id as usize].generation,
+                id,
+            })
+            .or_else(|| self.id_range.next().map(|id| Entity { generation: 0, id }))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.id_iter.len() + self.id_range.len();
+        (len, Some(len))
+    }
+}
+
+impl<'a> core::iter::ExactSizeIterator for ReserveEntitiesIterator<'a> {}
+
+#[derive(Debug, Default)]
+pub struct Entities {
+    pub meta: Vec<EntityMeta>,
+
+    // The `pending` and `free_cursor` fields describe three sets of Entity IDs
```

Comment:
Please use a doc comment. It would be useful when using `--document-private-items`.